d2/src/test/java/com/linkedin/d2/discovery/util/TestD2Config.java

```diff
@@ -392,6 +392,106 @@ public int getMaxPartitionId()
+  {
+    @SuppressWarnings("serial")
+    final Map<String,List<String>> clustersData = new HashMap<String,List<String>>()
+    {{
+      put("partitioned-cluster", Arrays.asList(new String[]{"partitioned-service-1", "partitioned-service-2"}));
+    }};
+
+    final Map<String, Object> partitionProperties = new HashMap<>();
+    Map<String, Object> customized = new HashMap<>();
+    List<String> classList = Arrays.asList("TestPartitionAccessor1!Settings", "TestPartitionAccessor2");
+    customized.put("partitionType", "CUSTOM");
+    customized.put("partitionCount", "10");
+    customized.put("partitionAccessorList", classList);
+    partitionProperties.put("partitionProperties", customized);
+
+    final PartitionAccessorRegistry registry = new PartitionAccessorRegistry()
+    {
+      final private Map<String, List<BasePartitionAccessor>> _registry = new HashMap<>();
+
+      @Override
+      public void register(String clusterName, BasePartitionAccessor accessor)
+      {
+        List<BasePartitionAccessor> accessors = _registry.computeIfAbsent(clusterName, k -> new ArrayList<>());
+        accessors.add(accessor);
+      }
+
+      @Override
+      public List<BasePartitionAccessor> getPartitionAccessors(String clusterName)
+      {
+        return _registry.get(clusterName);
+      }
+    };
+
+    class TestPartitionAccessor1 implements PartitionAccessor
+    {
+      @Override
+      public int getPartitionId(URI uri) throws PartitionAccessException
+      {
+        return testGetPartitionId(uri);
+      }
+
+      @Override
+      public int getMaxPartitionId()
+      {
+        return 10;
+      }
+
+      @Override
+      public boolean checkSupportable(String settings) {
+        return "TestPartitionAccessor1!Settings".equals(settings);
+      }
+    };
+
+    class TestPartitionAccessor2 implements PartitionAccessor
+    {
+      @Override
+      public int getPartitionId(URI uri) throws PartitionAccessException
+      {
+        return 8;
+      }
+
+      @Override
+      public int getMaxPartitionId()
+      {
+        return 10;
+      }
+    };
+
+    PartitionAccessor testAccessor1 = new TestPartitionAccessor1();
+    PartitionAccessor testAccessor2 = new TestPartitionAccessor2();
+
+    registry.register("partitioned-cluster", DefaultPartitionAccessor.getInstance());
+    registry.register("partitioned-cluster", testAccessor1);
+    registry.register("partitioned-cluster", testAccessor2);
+
+    D2ConfigTestUtil d2Conf = new D2ConfigTestUtil(clustersData, partitionProperties);
+
+    assertEquals(d2Conf.runDiscovery(_zkHosts), 0);
+
+    verifyPartitionProperties("partitioned-cluster", partitionProperties);
+
+    final ClusterProperties clusterprops = getClusterProperties(_zkclient, "partitioned-cluster" );
+
+    final PartitionAccessor accessor = PartitionAccessorFactory.getPartitionAccessor("partitioned-cluster",
+        registry, clusterprops.getPartitionProperties());
+
```

Comment:
you may also want to verify the class of the actual accessor.

Reply:
added